src/cascadia/QueryExtension/ExtensionPalette.cpp

```diff
@@ -0,0 +1,574 @@
+using namespace winrt::Windows::UI::Xaml;
+using namespace winrt::Windows::UI::Xaml::Controls;
+using namespace winrt::Windows::System;
+namespace WWH = ::winrt::Windows::Web::Http;
+namespace WSS = ::winrt::Windows::Storage::Streams;
+namespace WDJ = ::winrt::Windows::Data::Json;
+
+static constexpr std::wstring_view acceptedModel{ L"gpt-35-turbo" };
+static constexpr std::wstring_view acceptedSeverityLevel{ L"safe" };
+
+const std::wregex azureOpenAIEndpointRegex{ LR"(^https.*openai\.azure\.com)" };
+
+namespace winrt::Microsoft::Terminal::Query::Extension::implementation
+{
+    ExtensionPalette::ExtensionPalette()
+    {
+        InitializeComponent();
+
+        _clearAndInitializeMessages(nullptr, nullptr);
+        ControlName(RS_(L"ControlName"));
+        QueryBoxPlaceholderText(RS_(L"CurrentShell"));
+
+        auto disclaimerLinkText = Windows::UI::Xaml::Documents::Run();
+        disclaimerLinkText.Text(RS_(L"AIContentDisclaimerHyperlink"));
+        AIContentDisclaimerHyperlink().Inlines().Append(disclaimerLinkText);
+
+        auto learnMoreLinkText = Windows::UI::Xaml::Documents::Run();
+        learnMoreLinkText.Text(RS_(L"LearnMoreLink"));
+        LearnMoreLink().Inlines().Append(learnMoreLinkText);
+
+        _loadedRevoker = Loaded(winrt::auto_revoke, [this](auto /*s*/, auto /*e*/) {
+            // We have to add this in (on top of the visibility change handler below) because
+            // the first time the palette is invoked, we get a loaded event not a visibility event.
+
+            // Only let this succeed once.
+            _loadedRevoker.revoke();
+
+            _setFocusAndPlaceholderTextHelper();
+
+            // For the purposes of data collection, request the API key/endpoint *now*
+            _AIKeyAndEndpointRequestedHandlers(nullptr, nullptr);
+
+            TraceLoggingWrite(
+                g_hQueryExtensionProvider,
+                "QueryPaletteOpened",
+                TraceLoggingDescription("Event emitted when the AI chat is opened"),
+                TraceLoggingBoolean((!_AIKey.empty() && !_AIEndpoint.empty()), "AIKeyAndEndpointStored", "True if there is an AI key and an endpoint stored"),
+                TraceLoggingKeyword(MICROSOFT_KEYWORD_CRITICAL_DATA),
+                TelemetryPrivacyDataTag(PDT_ProductAndServiceUsage));
+        });
+
+        // Whatever is hosting us will enable us by setting our visibility to
+        // "Visible". When that happens, set focus to our query box.
+        RegisterPropertyChangedCallback(UIElement::VisibilityProperty(), [this](auto&&, auto&&) {
+            if (Visibility() == Visibility::Visible)
+            {
+                // Force immediate binding update so we can select an item
+                Bindings->Update();
+
+                _setFocusAndPlaceholderTextHelper();
+
+                // For the purposes of data collection, request the API key/endpoint *now*
+                _AIKeyAndEndpointRequestedHandlers(nullptr, nullptr);
```

Comment:
oh good gracious are we loading the keys from the vault every time it opens?

Reply:
yea :( making sure they hot reload sorta ;P